agb/src/interrupt.rs

```diff
@@ -21,37 +28,344 @@ pub enum Interrupt {
+                return;
+            }
+            c = p;
+        }
+    }
+}
+
+fn interrupt_to_root(interrupt: Interrupt) -> &'static InterruptRoot {
+    unsafe { &INTERRUPT_TABLE[interrupt as usize] }
+}
+
+fn get_interrupt_handle_root<'a>(
+    f: &'a dyn Fn(),
+    root: &InterruptRoot,
+) -> InterruptClosureBounded<'a> {
+    InterruptClosureBounded {
+        c: InterruptClosure {
+            closure: unsafe { core::mem::transmute(f as *const _) },
+            next: Cell::new(core::ptr::null()),
+            root: root as *const _,
+        },
+        _phantom: PhantomData,
+        _unpin: PhantomPinned,
+    }
+}
+
+pub fn get_interrupt_handle(
+    f: &(dyn Fn() + Send + Sync),
+    interrupt: Interrupt,
+) -> InterruptClosureBounded {
+    let root = interrupt_to_root(interrupt);
+
+    get_interrupt_handle_root(f, root)
+}
+
+pub fn add_interrupt<'a>(interrupt: Pin<&'a InterruptClosureBounded<'a>>) {
+    let root = unsafe { &*interrupt.c.root };
+    root.add();
+    let mut c = root.next.get();
+    if c.is_null() {
+        root.next.set((&interrupt.c) as *const _);
+        return;
+    }
+    loop {
+        let p = unsafe { &*c }.next.get();
+        if p.is_null() {
+            unsafe { &*c }.next.set((&interrupt.c) as *const _);
+            return;
+        }
+
+        c = p;
+    }
+}
+
+#[macro_export]
+macro_rules! add_interrupt_handler {
+    ($interrupt: expr, $handler: expr) => {
+        let a = $handler;
+        let a = $crate::interrupt::get_interrupt_handle(&a, $interrupt);
+        let a = unsafe { core::pin::Pin::new_unchecked(&a) };
+        $crate::interrupt::add_interrupt(a);
+    };
+}
+
+#[test_case]
+fn test_vblank_interrupt_handler(_gba: &mut crate::Gba) {
+    {
+        let counter = Mutex::new(0);
+        let counter_2 = Mutex::new(0);
+        add_interrupt_handler!(Interrupt::VBlank, || *counter.lock() += 1);
+        add_interrupt_handler!(Interrupt::VBlank, || *counter_2.lock() += 1);
+
+        let vblank = VBlank::get();
+
+        while *counter.lock() < 100 || *counter_2.lock() < 100 {
+            vblank.wait_for_vblank();
+        }
+    }
+
+    assert_eq!(
+        interrupt_to_root(Interrupt::VBlank).next.get(),
+        core::ptr::null(),
+        "expected the interrupt table for vblank to be empty"
+    );
+}
+
+#[test_case]
+fn test_interrupt_table_length(_gba: &mut crate::Gba) {
+    assert_eq!(
+        unsafe { INTERRUPT_TABLE.len() },
+        Interrupt::Gamepak as usize + 1,
+        "interrupt table should be able to store gamepak interrupt"
+    );
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+enum MutexState {
+    Unlocked,
+    Locked(bool),
+}
+
+pub struct Mutex<T> {
+    internal: UnsafeCell<T>,
+    state: UnsafeCell<MutexState>,
+}
+
+unsafe impl<T> Send for Mutex<T> {}
+unsafe impl<T> Sync for Mutex<T> {}
+
+impl<T> Mutex<T> {
+    pub fn lock(&self) -> MutexRef<T> {
+        let state = INTERRUPTS_ENABLED.get();
+        INTERRUPTS_ENABLED.set(0);
```

Comment:
There are a bunch of places where you do `INTERRUPTS_ENABLED.set(foo)`. Maybe extract that into a helper function?